GateAdventures/GateAdventures.cs

```diff
@@ -1,67 +1,71 @@
-﻿using System.Collections.Generic;
-using System.IO;
-using GateAdventures.Engine.Map;
+﻿using GateAdventures.Engine.Scenes;
 using Microsoft.Xna.Framework;
 using Microsoft.Xna.Framework.Graphics;
-using Microsoft.Xna.Framework.Input;
-using Newtonsoft.Json;
-using Newtonsoft.Json.Linq;
 
-namespace GateAdventures
+namespace GateAdventures;
+
+public class GateAdventures : Game
 {
-	public class GateAdventures : Game
-	{
-		private GraphicsDeviceManager _graphics;
-		private SpriteBatch _spriteBatch;
+	private readonly GraphicsDeviceManager _graphics;
+	private readonly SceneManager _sceneManager;
 
-		private List<TileMapRenderer> _maps;
 
-		public GateAdventures()
-		{
-			_graphics = new GraphicsDeviceManager(this);
-			Content.RootDirectory = "Content";
-			IsMouseVisible = true;
-		}
+	public GateAdventures()
+	{
+		_graphics = new GraphicsDeviceManager(this);
+		_sceneManager = new SceneManager(Services);
 
-		protected override void Initialize()
-		{
-			// TileMap width/height. We'll need to change this to 1920x1080
-			_graphics.PreferredBackBufferWidth = 960;
-			_graphics.PreferredBackBufferHeight = 960;
+		Content.RootDirectory = "Content";
+		IsMouseVisible = true;
+	}
 
-			_graphics.ApplyChanges();
-			base.Initialize();
-		}
+	protected override void Initialize()
+	{
+		_graphics.PreferredBackBufferWidth = 960;
+		_graphics.PreferredBackBufferHeight = 960;
+		_graphics.ApplyChanges();
 
-		protected override void LoadContent()
-		{
-			_spriteBatch = new SpriteBatch(GraphicsDevice);
+		SpriteBatch spriteBatch = new(GraphicsDevice);
 
-			// Load all configured TileMaps and their corresponding TileSets
-			_maps = TiledMapLoader.LoadAllMaps(Content, _graphics.GraphicsDevice);
-		}
+		Services.AddService(spriteBatch);
+		Services.AddService(_graphics);
+		Services.AddService(GraphicsDevice);
+		Services.AddService(_sceneManager);
+		Services.AddService(Content);
 
-		protected override void Update(GameTime gameTime)
-		{
-			if (GamePad.GetState(PlayerIndex.One).Buttons.Back == ButtonState.Pressed || Keyboard.GetState().IsKeyDown(Keys.Escape))
-				Exit();
+		_sceneManager.Start(new TileScene());
+		_sceneManager.StartOverlay(new HelloScene());
 
-			base.Update(gameTime);
-		}
+		base.Initialize();
+	}
 
-		protected override void Draw(GameTime gameTime)
-		{
-			GraphicsDevice.Clear(Color.CornflowerBlue);
-			_spriteBatch.Begin();
+	protected override void Update(GameTime gameTime)
+	{
+		IScene[] scenes = _sceneManager.Scenes;
+		int sceneCount = scenes.Length - 1;
```

Comment:
If `scenes` is empty, this would set `sceneCount` to `-1`, which doesn't make any sense. We should remove the `- 1` here, and make the for loop `i < sceneCount` imo

Reply:
That makes sense. I did not consider there ever being zero scenes